pkg/operator/ceph/controller/spec.go

```diff
@@ -388,15 +388,19 @@ func AppLabels(appName, namespace string) map[string]string {
 }
 
 // CephDaemonAppLabels returns pod labels common to all Rook-Ceph pods which may be useful for admins.
-// App name is the name of the application: e.g., 'rook-ceph-mon', 'rook-ceph-mgr', etc.
+// App name is the name of the application: e.g., 'rook-ceph-mon', 'rook-ceph-mgr', etc
 // Daemon type is the Ceph daemon type: "mon", "mgr", "osd", "mds", "rgw"
 // Daemon ID is the ID portion of the Ceph daemon name: "a" for "mon.a"; "c" for "mds.c"
-func CephDaemonAppLabels(appName, namespace, daemonType, daemonID string, includeNewLabels bool) map[string]string {
+// BinaryName is the name of the binary running in a container: "ceph-mon", "ceph-mgr", "ganesha.nfsd", etc
+// ResourceName is the resource metadata.name: rook-ceph-mon-a", "rook-ceph-osd-1", etc
+// ResourceKind is the CR type: "CephCluster", "CephFilesystem", etc
+func CephDaemonAppLabels(appName, namespace, daemonType, daemonID, binaryName, resourceName, resourceKind string, includeNewLabels bool) map[string]string {
 	labels := AppLabels(appName, namespace)
 
 	// New labels cannot be applied to match selectors during upgrade
 	if includeNewLabels {
 		labels[daemonTypeLabel] = daemonType
+		k8sutil.AddRecommendedLabels(labels, binaryName, resourceName, resourceKind, daemonType+"-"+daemonID)
 	}
 	labels[DaemonIDLabel] = daemonID
 	// Also report the daemon id keyed by its daemon type: "mon: a", "mds: c", etc.
```

Comment:
After some further thought, I think we can get rid of `binaryName` here. It seems a little less useful than using the daemon type, which is clearer at a glance. I think the full app name including the `rook-` prefix (`rook-ceph-<type>`) is too redundant, but I do think identifying the app as `ceph-` or not could be important, so let's instead build this up from `"ceph-"+daemonType`. This way, if we need to add these labels to other apps, we can prefix them with Rook.

Reply:
Updated